controllers/workspace/finalize.go

```diff
@@ -0,0 +1,213 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/controllers/workspace/provision"
+	"github.com/devfile/devworkspace-operator/internal/images"
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/go-logr/logr"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	pvcClaimMountPath   = "/tmp/devworkspaces/"
+	cleanupCommandFmt   = "rm -rf %s"
+	pvcCleanupFinalizer = "storage.controller.devfile.io"
+)
+
+var (
+	cleanupJobCompletions  = int32(1)
+	cleanupJobBackoffLimit = int32(3)
+)
+
+func (r *DevWorkspaceReconciler) setFinalizer(ctx context.Context, workspace *v1alpha2.DevWorkspace) (ok bool, err error) {
+	if !isFinalizerNecessary(workspace) || hasFinalizer(workspace) {
+		return true, nil
+	}
+	workspace.SetFinalizers(append(workspace.Finalizers, pvcCleanupFinalizer))
+	return false, r.Update(ctx, workspace)
+}
+
+func (r *DevWorkspaceReconciler) finalize(ctx context.Context, log logr.Logger, workspace *v1alpha2.DevWorkspace) error {
+	if !hasFinalizer(workspace) {
+		return nil
+	}
+	// Need to make sure Deployment is cleaned up before starting job to avoid mounting issues for RWO PVCs
+	wait, err := provision.DeleteWorkspaceDeployment(ctx, workspace, r.Client)
+	if err != nil {
+		return err
+	}
+	if wait {
+		return nil
+	}
+	specJob, err := r.getSpecCleanupJob(workspace)
+	if err != nil {
+		return err
+	}
+	clusterJob, err := r.getClusterCleanupJob(ctx, workspace)
+	if err != nil {
+		return err
+	}
+	if clusterJob == nil {
+		err := r.Create(ctx, specJob)
+		if err != nil && !k8sErrors.IsAlreadyExists(err) {
+			return err
+		}
+		return nil
+	}
+	if !equality.Semantic.DeepDerivative(specJob.Spec, clusterJob.Spec) {
+		err := r.Patch(ctx, specJob, client.Merge)
```

Comment:
please double check if we're able to patch job, I may be wrong but I suppose we must recreate job instead.

Reply:
It's some fields are patchable, others seem to not be:
- The standard rules apply for the pod template (can only change a few of the fields)
- I can update restartCount but not completions
- etc.

It's probably safer to just delete the job and recreate it.

Reply:
Set it up to delete Job if it doesn't match spec.